Modules type

dvdherron opened the pull request:
![](https://source.unsplash.com/featured/?cute,animal&tked)
<!-- Swap `CHANGE_ME` for a random string to get a different image -->



## Description
Upgrading the `type` module  with Sass module usage. 

## Steps to test/reproduce

See files in `./sass/type/`

_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._


## Show me
_Provide screenshots/animated gifs/videos if necessary._

dvdherron commented:
![](https://github.trello.services/images/mini-trello-icon.png) [I want Accoutrement on Sass Modules](https://trello.com/c/9tyINy2o/23-i-want-accoutrement-on-sass-modules)

mirisuzanne commented:
![](https://github.trello.services/images/mini-trello-icon.png) [I want Accoutrement on Sass Modules](https://trello.com/c/9tyINy2o/23-i-want-accoutrement-on-sass-modules)

dvdherron commented:
@mirisuzanne I'm working through updating this module and fixing the tests. 

Tests are passing for `config`, `a11y`, and `pseudo-elements`.  I still need to work on `font-vars`, `fonts` and `helpers`.

Three tests are failing in `normalize`. (I've commented them out).  Can you tell if there's something wrong in the tests or in `type/normalize` that's causing them to fail?

mirisuzanne commented on sass/type/_normalize.scss:
```diff
@@ -129,14 +129,6 @@
-      string.unquote($name)
-    );
-  }
-
```
It may have been a mistake to remove this… but at the time it seemed redundant, and didn't have any tests.

dvdherron replied:
hmm. without these lines, the first two tests in `Fonts` fail due to the extra quotes:
![image](https://user-images.githubusercontent.com/41588129/141302318-e8991706-4078-4238-97b9-b9ff8ca03926.png)

mirisuzanne replied:
super weird that they didn't fail for me. but good catch - @dvdherron can you add these lines back in? (since this is used a couple places, it could also become a function if you wanted)

dvdherron replied:
yeah, strange. And I will look into turning this into a function . . .

mirisuzanne commented on test/type/_normalize.scss:
```diff
@@ -6,25 +6,26 @@
 
-// Test Font Normalization
-// =======================
+$_global: config.$fonts;
+config.$fonts: data.$fonts;
```
All these functions take a `$source` argument. If it's not set explicitly, they look up `config.$fonts`. We need to make sure they find the actual source map we're trying to test.

dvdherron replied:
Ah, ok. I saw you doing this in the test for `colors` and now that makes sense.

mirisuzanne commented on test/type/_normalize.scss:
```diff
@@ -6,25 +6,26 @@
       @include assert-equal(
         map.get($test, 'stack'),
-        ('TestFont', Georgia, serif)
+        (TestFont, Georgia, serif)
```
The quoting and unquoting of names in a font stack is important. Output should only have quotes when the font name in question has spaces.

dvdherron commented on sass/type/_helpers.scss:
```diff
@@ -147,19 +144,15 @@ $_a_FONT-FORMATS: (
   // stylelint-disable-next-line function-url-quotes
-  $path: _a_font-url($path, $format, $iefix: true);
+  $path: font-url($path, $format, $iefix: true);
   @return font-url($path) $format-name;
```
@mirisuzanne As you know, the second `font-url` function here is defined in the `fonts` module. if I use that module here and try to call it with `fonts.font-url` I get a module loop error:

```
Error: Module loop: this module is already being loaded.
  ┌──> sass/type/_helpers.scss
5 │ @use './fonts';
  │ ^^^^^^^^^^^^^^ new load
  ╵
  ┌──> test/type/_fonts.scss
3 │ @use '../../../sass/type/fonts';
  │ ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ original load
```
Would the best way to go about this be to relocate/rename the `font-url` function in `fonts`?

mirisuzanne replied:
For sure we should rename one or both of the font-url functions, to clarify what they do. But then, yeah, we can't have both modules relying on the other - so _something_ has to relocate. Maybe it works to move the `fonts` one into helpers? Since that's the public one, it's also the one that should get priority for keeping the name.

dvdherron replied:
@mirisuzanne 
I ended up moving the `fonts` version into `helpers` and renamed the `helpers` version to `format-url`.

All `type` tests are passing now except for one in `fonts`.

 ```
Import Webfonts [mixin]
         Font-face imports for all local fonts & variants:

```

If I remove the two FranklinGothic `@font-face` blocks, the test passes. I think this means that the test is looking for those two fonts in the global `config.$fonts map`. 

When I add FranklinGothic to the `test-data` font map, it makes the first test in type/config fail
![image](https://user-images.githubusercontent.com/41588129/142269732-dadff4e4-75b8-4a73-b081-29a9ded10713.png)

I think that tests assumes that Franklin isn't in the global fonts map, right? 

Should I just remove to two `@font-face` blocks for `FranklinGothic` in the `Import Webfonts [mixin]` test? 

We can schedule a time to walk through this if this is hard to follow :)

dvdherron commented on sass/type/_font-vars.scss:
```diff
@@ -25,12 +31,11 @@
 ///    )
 /// );
 /// $font-var-prefix: 'prefix_';
```
```suggestion
/// tools.$font-var-prefix: 'prefix_';
```

dvdherron commented on sass/type/_font-vars.scss:
```diff
@@ -25,12 +31,11 @@
+/// html { @include tools.font--('my-font') }
 ///
 /// @type string
 $font-var-prefix: 'font-' !default;
```
```suggestion
tools.$font-var-prefix: 'font-' !default;
```

mirisuzanne replied:
This is the initial definition of the variable, so should not be prefixed.

dvdherron commented on sass/type/_index.scss:
```diff
@@ -7,35 +7,23 @@
-
+@forward 'config';
+@forward 'helpers';
+@forward 'normalize';
```
```suggestion

```

dvdherron commented on sass/type/_font-vars.scss:
```diff
@@ -192,22 +191,29 @@ $font-var-prefix: 'font-' !default;
-    $normal: _a_normalize-font($normal, $key, $source);
-    $has: map-get($normal, $value);
+    $normal: api.get($source, $key);
+    $normal: normalize.normalize-font($normal, $key, $source);
```
```suggestion
    $normal: normalize.font($normal, $key, $source);
```

dvdherron commented on sass/type/_fonts.scss:
```diff
@@ -30,45 +38,12 @@
-  $normal: _a_normalize-font($data, $font, $source);
+@function font($font, $source: config.$fonts) {
+  $data: api.get($source, $font);
+  $normal: normalize.normalize-font($data, $font, $source);
```
```suggestion
  $normal: normalize.font($data, $font, $source);
```

dvdherron commented on sass/type/_fonts.scss:
```diff
@@ -168,35 +135,38 @@
+  $variants: helpers.font-get-variants($font);
+  $name: map.get($font, 'name');
 
   @if (length($variants) > 0) {
```
```suggestion
  @if (list.length($variants) > 0) {
```

dvdherron commented on sass/type/_index.scss:
```diff
@@ -7,35 +7,23 @@
-@import 'pseudo-elements';
-
+@forward 'config';
+@forward 'helpers';
```
```suggestion

```

dvdherron commented on sass/type/_helpers.scss:
```diff
@@ -363,8 +361,36 @@ $_a_FONT-FORMATS: (
   }
 }
+
+// Font-URL [function]
```
public, so should be forwarded. use @forward show ...

dvdherron commented on sass/type/_helpers.scss:
```diff
@@ -363,8 +361,36 @@ $_a_FONT-FORMATS: (
+///
+/// @group type-fonts
+/// @example scss
+///   $font-path: '../fonts/';
```
```suggestion
///   tools.$font-path: '../fonts/';
```

dvdherron commented on sass/type/_helpers.scss:
```diff
@@ -363,8 +361,36 @@ $_a_FONT-FORMATS: (
+/// @group type-fonts
+/// @example scss
+///   $font-path: '../fonts/';
+///   $src: font-url('sans/my-sans-font.woff2');
```
```suggestion
///   $src: tools.font-url('sans/my-sans-font.woff2');
```

dvdherron commented on sass/type/_normalize.scss:
```diff
@@ -24,42 +34,59 @@
-  $key: null,
-  $source: $fonts
-) {
+@function normalize-font($font, $key: null, $source: config.$fonts) {
```
```suggestion
@function normalize.font($font, $key: null, $source: config.$fonts) {
```

dvdherron commented on sass/type/_normalize.scss:
```diff
@@ -24,42 +34,59 @@
   // Normalize name…
-  $name: _a_normalize-font-name($font, $key, $source);
-  $font: if($name, map-merge($font, ('name': $name)), $font);
+  $name: normalize-font-name($font, $key, $source);
```
```suggestion
  $name: font-name($font, $key, $source);
```

dvdherron commented on sass/type/_normalize.scss:
```diff
@@ -24,42 +34,59 @@
   // Normalize the font stack…
-  $stack: _a_normalize-font-stack($font);
-  $font: if($stack, map-merge($font, ('stack': $stack)), $font);
+  $stack: normalize-font-stack($font);
```
```suggestion
  $stack: font-stack($font);
```

dvdherron commented:
@mirisuzanne @stacyk 

I cleaned up based on the review:
- made `normalize` and `helper` private
- renamed the functions in normalize so that they would not be double referenced. 
- and fixed the missing module references 

It seems like I'm missing loading a module somewhere though. When I remove the type files from the `IGNORE_SASS` list in the gulpfile, I get errors about undefined mixins and variables in the examples:
```
» [WARNING] Error compiling @example scss: 
Undefined mixin.
   ╷
3  │ ┌ @include tools.import-font-face(
4  │ │   $name: 'Franklin Gothic',
5  │ │   $weight: 'normal',
6  │ │   $style: 'normal',
7  │ │   $data: (
8  │ │     'local': 'Franklin Gothic' ('franklin-gothic'),
9  │ │     'woff2': 'path/to/franklin-bolditalic-font.woff2',
10 │ │     'ttf': 'path/to/franklin-bolditalic-font.ttf',
11 │ │   )
12 │ │ );
```
   │ └─^

I'll take a closer look at what's happening here tomorrow.

mirisuzanne commented on sass/type/_index.scss:
````diff
@@ -68,22 +57,22 @@
 /// ```scss
 ///  // Import one font by configuration key
-///  @include font-face('body');
+///  @include tools.font-face('body');
````
Don't prefix markdown examples? (see multiple on this page)

mirisuzanne reviewed (commented):
Looking pretty good, just needs some final cleanup?
